add session transcript for browser api (#374)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

* Rules for the `presentation_definition` Authorization Request parameter.
* Rules for the `presentation_submission` Authorization Response parameter.
* Wallet invocation using the `mdoc-openid4vp://` custom URI scheme.

> * Required Wallet and Verifier Metadata parameters and their values when OpenID4VP is used with the `mdoc-oid4vp://` custom URI scheme.

Reviewer: ```suggestion
* Required Wallet and Verifier Metadata parameters and their values.
* The `SessionTranscript` and `Handover` CBOR structure when OpenID4VP is used with the `mdoc-oid4vp://` custom URI scheme. Also see (#non-dc-api-invocation).
```

18013-7 does define `SessionTranscript` and `Handover` CBOR structure when OpenID4VP is used with the `mdoc-oid4vp://` custom URI scheme, so I would still mention it

Author: That is fine but let's make the other item not `mdoc-oid4vp://` specific.

```suggestion
* Required Wallet and Verifier Metadata parameters and their values when OpenID4VP is used with the `mdoc-oid4vp://` custom URI scheme.
The `SessionTranscript` and `Handover` CBOR structure when the invocation does not use the DC API. Also see (#non-dc-api-invocation).
```

Author: Does that work for you @Sakurann ?

Author: Committed my suggestion as agreed on the DCP call.

See ISO/IEC TS 18013-7 Annex B [@ISO.18013-7] and ISO/IEC 23220-4 Annex C [@ISO.23220-4] for the latest examples on how to use the `presentation_submission` parameter and how to generate the Authorizaton Response for presenting Credentials in the mdoc format.

### `Handover` and `SessionTranscript` Definitions

#### Invocation via the Digital Credentials API

If the presentation request is invoked using the Digital Credentials API, the `SessionTranscript` CBOR structure as defined in Section 9.1.5.1 in [@ISO.18013-5] MUST be used with the following changes:

* `DeviceEngagementBytes` MUST be `null`.
* `EReaderKeyBytes` MUST be `null`.
* `Handover` MUST be the `OID4VPDCAPIHandover` CBOR structure as defined below.

```cddl
OID4VPDCAPIHandover = [
  "OID4VPDCAPIHandover", ; A fixed identifier for this handover type
  OID4VPDCAPIHandoverInfoHash ; A cryptographic hash of OID4VPDCAPIHandoverInfo
]

> OID4VPDCAPIHandoverInfoHash = bstr  ; sha-256 hash of OID4VPDCAPIHandoverInfo

Reviewer: ```suggestion

OID4VPDCAPIHandoverInfoHash = bstr  ; sha-256 hash of OID4VPDCAPIHandoverInfo

```

can we have empty lines to improve readability?

OID4VPDCAPIHandoverInfo = [
  origin,
  client_id,
  nonce
] ; Array containing handover parameters

> client_id = tstr  ; UTF-8 encoded string

Reviewer: ```suggestion

client_id = tstr  ; UTF-8 encoded string
```
same as above, re readability

origin = tstr    ; UTF-8 encoded string
nonce = tstr     ; UTF-8 encoded string
```

The `OID4VPDCAPIHandover` structure has the following elements:

> - The first element MUST be the fixed UTF-8 encoded string `"OID4VPDCAPIHandover"`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.

Reviewer: ```suggestion
- The first element MUST be the fixed UTF-8 encoded string `OID4VPDCAPIHandover`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.
```

i don't remember how 18013-5 and 18013-7 is written. this is probably less confusing since this text is outside CDDL definition

> - The second element MUST be the `OID4VPDCAPIHandoverInfoHash`, represented as a CBOR byte string which encodes the sha-256 hash of the `OID4VPDCAPIHandoverInfo` CBOR array.
> - The `OID4VPDCAPIHandoverInfo` has the following elements:

Reviewer 2: ```suggestion
- The second element MUST be the `OID4VPDCAPIHandoverHash`, represented as a CBOR byte string which encodes the sha-256 hash of the `OID4VPDCAPIHandover` CBOR array.
- The `OID4VPDCAPIHandover` has the following elements:
```

> - The second element MUST be the `OID4VPDCAPIHandoverInfoHash`, represented as a UTF-8 string. This string encodes the base64-encoded (as defined in Section 4 of [@!RFC4648] -- not base64url-encoded) sha-256 hash of the `OID4VPDCAPIHandoverInfo` CBOR array, e.g., `H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO`.

Reviewer 3: Is there any benefit in encoding w/ base64url here or should we just use the raw Hash?
```cddl
ID4VPDCAPIHandover = [
  "OID4VPDCAPIHandover", ; A fixed identifier for this handover type
  OID4VPDCAPIHandoverInfoHash ; A cryptographic hash of OID4VPDCAPIHandoverInfo
]
OID4VPDCAPIHandoverInfoHash = bstr ; sha-256 hash of OID4VPDCAPIHandoverInfo
```

Author: No real benefit, just consistency reasons. I thought having UTF-8 strings consistently would reduce some encoding issues in implementations. I updated the PR to use bstr instead of tstr.

Author: Hash is using the raw hash, no longer W3C SRI.

  - The first element MUST be the UTF-8 encoded string representing the origin of the request as described in (#dc_api_request).

>   - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter.

Reviewer 4: Shouldn't that be the effective Client Identifier (as defined in (#dc_api_request)? I'm asking since there is no `client_id` parameter in case of an unsigned DC API request.

Author: Yes, please review again. PR got shortened and previously had this info.

>   - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter.

Reviewer 4: ```suggestion
  - The second element MUST be the UTF-8 encoded string value of the effective Client Identifier as defined in (#dc_api_request).
```

> - The first element MUST be the fixed UTF-8 encoded string `"OID4VPDCAPIHandover"`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.
> - The second element MUST be the `OID4VPDCAPIHandoverInfoHash`, represented as a CBOR byte string which encodes the sha-256 hash of the `OID4VPDCAPIHandoverInfo` CBOR array.
> - The `OID4VPDCAPIHandoverInfo` has the following elements:
>   - The first element MUST be the UTF-8 encoded string representing the origin of the request as described in (#dc_api_request).
>   - The second element MUST be the UTF-8 encoded string value of the effective Client Identifier as defined in (#dc_api_request).
>   - The third element MUST be the UTF-8 encoded string value of the `nonce` request parameter.

Reviewer: ```suggestion
* The first element MUST be the fixed UTF-8 encoded string `"OID4VPDCAPIHandover"`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.
* The second element MUST be the `OID4VPDCAPIHandoverInfoHash`, represented as a CBOR byte string which encodes the sha-256 hash of the `OID4VPDCAPIHandoverInfo` CBOR array.
* The `OID4VPDCAPIHandoverInfo` has the following elements:
  * The first element MUST be the UTF-8 encoded string representing the origin of the request as described in (#dc_api_request).
  * The second element MUST be the UTF-8 encoded string value of the effective Client Identifier as defined in (#dc_api_request).
  * The third element MUST be the UTF-8 encoded string value of the `nonce` request parameter.
```

because someone corrected my PR like this previously...

>   - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests, the value for the `client_id` MUST be derived from the `origin` value as defined in (#dc_api_request), e.g., `web-origin:https://verifier.example.com`.

Reviewer 5: Why are we using the term “audience binding”? I believe either we clarify its meaning here or we remove it altogether.

Author: Unfortunately, I wasn't on the call when this was discussed but here is the summary of why audience binding is included: https://github.com/openid/oid4vc-haip/issues/135#issuecomment-2549616030

The way I interpret it—but others, please feel free to correct me—is that the `client_id` represents the audience, which is the requester of the credential. For unsigned requests, the audience and the `origin` are generally the same. However, for signed requests, the `origin` value can differ from the audience identifier. In such cases, the audience must also verify that the `origin` is one of their expected origins. Essentially, `client_id` is used for audience binding, while `origin` is used for origin binding, though for unsigned requests, both are more or less the same.

Author: As agreed on the DCP call, I removed the reasoning.

See ISO/IEC TS 18013-7 Annex B [@ISO.18013-7] and ISO/IEC 23220-4 Annex C [@ISO.23220-4] for the latest examples on how to use the `presentation_submission` parameter and how to generate the Authorizaton Response for presenting Credentials in the mdoc format.

### `Handover` and `SessionTranscript` Definitions

#### Invocation via the Digital Credentials API

If the presentation request is invoked using the Digital Credentials API, the `SessionTranscript` CBOR structure as defined in Section 9.1.5.1 in [@ISO.18013-5] MUST be used with the following changes:

* `DeviceEngagementBytes` MUST be `null`.
* `EReaderKeyBytes` MUST be `null`.
* `Handover` MUST be the `OpenID4VPDCAPIHandover` CBOR structure as defined below.

```cddl
OpenID4VPDCAPIHandover = [
  "OpenID4VPDCAPIHandover", ; A fixed identifier for this handover type
  OpenID4VPDCAPIHandoverInfoHash ; A cryptographic hash of OpenID4VPDCAPIHandoverInfo
]

OpenID4VPDCAPIHandoverInfoHash = bstr  ; sha-256 hash of OpenID4VPDCAPIHandoverInfo

> OpenID4VPDCAPIHandoverInfoHash = bstr  ; sha-256 hash of OpenID4VPDCAPIHandoverInfo
> 

Reviewer 3: One of the commits somehow duplicated the line?
```suggestion
```

See ISO/IEC TS 18013-7 Annex B [@ISO.18013-7] and ISO/IEC 23220-4 Annex C [@ISO.23220-4] for the latest examples on how to use the `presentation_submission` parameter and how to generate the Authorizaton Response for presenting Credentials in the mdoc format.

### `Handover` and `SessionTranscript` Definitions

> #### DC API

Reviewer 6: ```suggestion
#### Invocation via the Digital Credentials API
```

> If the DC API is used, the `SessionTranscript` CBOR structure as defined in Section 9.1.5.1 in [ISO.18013-5] MUST be used with the following changes:

Reviewer 6: ```suggestion
If the presentation request is invoked using the Digital Credentials API, the `SessionTranscript` CBOR structure as defined in Section 9.1.5.1 in [ISO.18013-5] MUST be used with the following changes:
```

```cddl
OID4VPDCAPIHandover = [
  "OID4VPDCAPIHandover", ; A fixed identifier for this handover type
  OID4VPDCAPIHandoverInfoHash ; Integrity hash of OID4VPDCAPIHandoverInfo
]
client_id = tstr  ; UTF-8 encoded string
origin = tstr    ; UTF-8 encoded string
nonce = tstr     ; UTF-8 encoded string
OID4VPDCAPIHandoverInfo = [ origin, client_id, nonce ] ; Array containing handover parameters
OID4VPDCAPIHandoverInfoHash = tstr  ; UTF-8 encoded string for the integrity hash of OID4VPDCAPIHandoverInfo
```

`OID4VPDCAPIHandover` CBOR structure elements:

- The first element MUST be the fixed UTF-8 encoded string `"OID4VPDCAPIHandover"`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.
- The second element MUST be the `OID4VPDCAPIHandoverInfoHash` encoded as a UTF-8 string representing the integrity hash of the `OID4VPDCAPIHandoverInfo` CBOR array.
- The value of `OID4VPDCAPIHandoverInfoHash` MUST comply with the W3C Subresource integrity format as defined in [!W3C.SRI] (see below), e.g., `sha256-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO`.
- The `OID4VPDCAPIHandoverInfo` has the following elements:

>   - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be obtained from the web platform or app platform being used.

Reviewer 6: ```suggestion
  - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be obtained from the Digital Credentials API client (the Web Platform or App Platform hosting the wallet).
```

>   - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be obtained from the web platform or app platform being used.

Reviewer 6: ```suggestion
  - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be obtained from the Digital Credentials API request.
```

Reviewer 6: @leecam thoughts?

Reviewer 7: I think this wording could be confusing - the origin we want is coming from the platform. By the time we're into the platform layers and the browser/platform has added something I don't think it's accurate to describe it as a "Digital Credentials API request".

Maybe this is better? Not sure but maybe someone can improve upon it :) 

```suggestion
  - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be the one the web platform or app platform asserted the request was made by.
```

  - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests the value for the `client_id` MUST be derived from the `origin` value.
  - The third element MUST be the UTF-8 encoded string value of the `nonce` request parameter to enable session binding.

#### Non-DC API

> If the DC API is not used, the rules for generating the `SessionTranscript` and `Handover` CBOR structure are specified in ISO/IEC 18013-5 [ISO.18013-5] and ISO/IEC 23220-4 [ISO.23220-4].

Reviewer 6: ```suggestion
If the presentation request is invoked via other methods, the rules for generating the `SessionTranscript` and `Handover` CBOR structure are specified in ISO/IEC 18013-5 [ISO.18013-5] and ISO/IEC 23220-4 [ISO.23220-4].
```

> #### Non-DC API
> 
> If the DC API is not used, the rules for generating the `SessionTranscript` and `Handover` CBOR structure are specified in ISO/IEC 18013-5 [ISO.18013-5] and ISO/IEC 23220-4 [ISO.23220-4].

Reviewer: ```suggestion
#### Invocation without using the Digital Credentials API {#non-dc-api-invocation}

If the Digital Credentials API is not used to invoke the wallet, the rules for generating the `SessionTranscript` and `Handover` CBOR structure are specified in ISO/IEC 18013-5 [ISO.18013-5] and ISO/IEC 23220-4 [ISO.23220-4].
```

nonce = tstr     ; UTF-8 encoded string
OID4VPDCAPIHandoverInfo = [ origin, client_id, nonce ] ; Array containing handover parameters
OID4VPDCAPIHandoverInfoHash = tstr  ; UTF-8 encoded string for the integrity hash of OID4VPDCAPIHandoverInfo
```

`OID4VPDCAPIHandover` CBOR structure elements:

- The first element MUST be the fixed UTF-8 encoded string `"OID4VPDCAPIHandover"`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.
- The second element MUST be the `OID4VPDCAPIHandoverInfoHash` encoded as a UTF-8 string representing the integrity hash of the `OID4VPDCAPIHandoverInfo` CBOR array.
- The value of `OID4VPDCAPIHandoverInfoHash` MUST comply with the W3C Subresource integrity format as defined in [!W3C.SRI] (see below), e.g., `sha256-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO`.
- The `OID4VPDCAPIHandoverInfo` has the following elements:
  - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be the one the web platform or app platform asserted the request was made by.

>   - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests the value for the `client_id` MUST be derived from the `origin` value.

Reviewer 6: ```suggestion
  - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests, the value for the `client_id` MUST be derived from the `origin` value.
```

nit

See ISO/IEC TS 18013-7 Annex B [@ISO.18013-7] and ISO/IEC 23220-4 Annex C [@ISO.23220-4] for the latest examples on how to use the `presentation_submission` parameter and how to generate the Authorizaton Response for presenting Credentials in the mdoc format.

### `Handover` and `SessionTranscript` Definitions

#### Invocation via the Digital Credentials API

If the presentation request is invoked using the Digital Credentials API, the `SessionTranscript` CBOR structure as defined in Section 9.1.5.1 in [ISO.18013-5] MUST be used with the following changes:

* `DeviceEngagementBytes` MUST be `null`.
* `EReaderKeyBytes` MUST be `null`.

> * `Handover` MUST be the `OID4VPDCAPIHandover` CBOR structure as defined below.

Reviewer 2: As we had the naming discussion recently, do we want these values to be OPENID4VP... ?

Author: Not sure why anyone would confuse CBOR structure names with OIDs? But I can update the PR to use `OpenID4VPDigitalCredentialsAPIHandover` if this works for you?

Reviewer: `OpenID4VPDCAPIHandover` is probably sufficient.

> Not sure why anyone would confuse CBOR structure names with OIDs?
I am 100% with you Oliver, but this has came up enough times, unfortunately.......

```cddl
OID4VPDCAPIHandover = [
  "OID4VPDCAPIHandover", ; A fixed identifier for this handover type

>   OID4VPDCAPIHandoverInfoHash ; Integrity hash of OID4VPDCAPIHandoverInfo

Reviewer 2: What does Integrity mean in this context? It seems confusing to me. If Hash alone isn't enough, I would rather write cryptographic hash?

Author: @paulbastian I removed "integrity" since it is just the hash.

]

> client_id = tstr  ; UTF-8 encoded string
> origin = tstr    ; UTF-8 encoded string
> nonce = tstr     ; UTF-8 encoded string
> OID4VPDCAPIHandoverInfo = [
>   origin,
>   client_id,
>   nonce
> ] ; Array containing handover parameters
> OID4VPDCAPIHandoverInfoHash = tstr  ; UTF-8 encoded string for the integrity hash of OID4VPDCAPIHandoverInfo

Reviewer 2: Minor nitpicking, but ISO usually has the order the other way around, going from top level to bottom consistently
```suggestion
OID4VPDCAPIHandoverInfo = [
  origin,
  client_id,
  nonce
] ; Array containing handover parameters
OID4VPDCAPIHandoverInfoHash = tstr  ; UTF-8 encoded string for the integrity hash of OID4VPDCAPIHandoverInfo
client_id = tstr  ; UTF-8 encoded string
origin = tstr    ; UTF-8 encoded string
nonce = tstr     ; UTF-8 encoded string
```

```

> `OID4VPDCAPIHandover` CBOR structure elements:

Reviewer 2: ```suggestion
The `OID4VPDCAPIHandover` structure has the following elements:
```

- The first element MUST be the fixed UTF-8 encoded string `"OID4VPDCAPIHandover"`. This serves as a unique identifier for the handover structure to prevent misinterpretation or confusion.
- The second element MUST be the `OID4VPDCAPIHandoverInfoHash` encoded as a UTF-8 string representing the integrity hash of the `OID4VPDCAPIHandoverInfo` CBOR array.

> - The value of `OID4VPDCAPIHandoverInfoHash` MUST comply with the W3C Subresource integrity format as defined in [!W3C.SRI] (see below), e.g., `sha256-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO`.
> - The `OID4VPDCAPIHandoverInfo` has the following elements:
>   - The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be the one the web platform or app platform asserted the request was made by.
>   - The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests, the value for the `client_id` MUST be derived from the `origin` value.
>   - The third element MUST be the UTF-8 encoded string value of the `nonce` request parameter to enable session binding.

Reviewer 2: ```suggestion
  - The value of `OID4VPDCAPIHandoverInfoHash` MUST comply with the W3C Subresource integrity format as defined in [!W3C.SRI] (see below), e.g., `sha256-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO`.

The `OID4VPDCAPIHandoverInfo` has the following elements:
- The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be the one the web platform or app platform asserted the request was made by.
- The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests, the value for the `client_id` MUST be derived from the `origin` value.
- The third element MUST be the UTF-8 encoded string value of the `nonce` request parameter to enable session binding.
```

Reviewer 5: Should we refer to the section in the specification that explains how the `client_id` must be derived from the `origin` value for unsigned requests.

Author: @hlozi @paulbastian would the following work for you?

```suggestion
  - The value of `OID4VPDCAPIHandoverInfoHash` MUST comply with the W3C Subresource integrity format as defined in [!W3C.SRI] (see below), e.g., `sha256-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO`.
The `OID4VPDCAPIHandoverInfo` structure has the following elements:
- The first element MUST be the UTF-8 encoded string representing the `origin` of the Verifier to protect against MITM attacks. The value for `origin` MUST be the one the web platform or app platform asserted the request was made by.
- The second element MUST be the UTF-8 encoded string value of the `client_id` request parameter if the request was signed for audience binding. For unsigned requests, the `origin` value MUST be used for the `client_id` value.
- The third element MUST be the UTF-8 encoded string value of the `nonce` request parameter to enable session binding.
```

Reviewer 7: > For unsigned requests, the `origin` value MUST be used for the `client_id` value.

This seems like it implies `client_id == "https://whatever"`, but it should be `client_id == "web-origin:https://whatever"`

I suggest we just refer to https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html#appendix-A.2 instead of trying to add normative text that duplicates what is already in that section.

Author: I updated the PR accordingly. Please check again @jogu .
